[inductor] Bypass FX graph cache when we have HigherOrderOperators

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123325

Summary: The initial motivation was to avoid caching when we have triton higher order ops, but it's probably safer to avoid the cache for all higher order ops and allow/implement if/when we find it necessary.

Test Plan: Unit test cribbed from: https://docs-preview.pytorch.org/pytorch/tutorials/2783/recipes/torch_compile_user_defined_triton_kernel_tutorial.html?highlight=triton

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

masnesral committed:
[inductor] Bypas FX graph cache when we have HigherOrderOperators

Summary: The initial motivation was to avoid caching when we have triton higher order ops, but it's probably safer to avoid the cache for all higher order ops and allow/implement if/when we find it necessary.

Test Plan: Unit test cribbed from: https://docs-preview.pytorch.org/pytorch/tutorials/2783/recipes/torch_compile_user_defined_triton_kernel_tutorial.html?highlight=triton

[ghstack-poisoned]

masnesral committed:
lint on "[inductor] Bypass FX graph cache when we have HigherOrderOperators"


Summary: The initial motivation was to avoid caching when we have triton higher order ops, but it's probably safer to avoid the cache for all higher order ops and allow/implement if/when we find it necessary.

Test Plan: Unit test cribbed from: https://docs-preview.pytorch.org/pytorch/tutorials/2783/recipes/torch_compile_user_defined_triton_kernel_tutorial.html?highlight=triton

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Bypass FX graph cache when we have HigherOrderOperators"


Summary: The initial motivation was to avoid caching when we have triton higher order ops, but it's probably safer to avoid the cache for all higher order ops and allow/implement if/when we find it necessary.

Test Plan: Unit test cribbed from: https://docs-preview.pytorch.org/pytorch/tutorials/2783/recipes/torch_compile_user_defined_triton_kernel_tutorial.html?highlight=triton

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
rebase on "[inductor] Bypass FX graph cache when we have HigherOrderOperators"


Summary: The initial motivation was to avoid caching when we have triton higher order ops, but it's probably safer to avoid the cache for all higher order ops and allow/implement if/when we find it necessary.

Test Plan: Unit test cribbed from: https://docs-preview.pytorch.org/pytorch/tutorials/2783/recipes/torch_compile_user_defined_triton_kernel_tutorial.html?highlight=triton

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Bypass FX graph cache when we have HigherOrderOperators"


Summary: The initial motivation was to avoid caching when we have triton higher order ops, but it's probably safer to avoid the cache for all higher order ops and allow/implement if/when we find it necessary.

Test Plan: Unit test cribbed from: https://docs-preview.pytorch.org/pytorch/tutorials/2783/recipes/torch_compile_user_defined_triton_kernel_tutorial.html?highlight=triton

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen commented on test/inductor/test_codecache.py:
```diff
@@ -287,6 +291,50 @@ def fn2(x):
+        """
+
+        @triton.jit
+        def add_kernel(
```
no need to redefine this, import it from torch/testing/_internal/triton_utils.py

oulgen commented on torch/_inductor/codecache.py:
```diff
@@ -821,21 +821,27 @@ def _save_graph(
             log.debug("fx graph cache no shape env")
             raise BypassFxGraphCache()
 
+        # HigherOrderOperators should be handled on a case-by-case basis.
+        # Currently, we just skip caching if we have any.
```
should we be skipping all of them by default? i assume cond/map etc should just work out of the box

masnesral replied:
This was @eellison's suggestion. Elias, what do you think? What is the set of higher order ops? Is it currently just triton + cond/map, or are there others as well? And if there are others, is there a criteria to distinguish which we expect to work vs. not?

oulgen replied:
A subset of all higher order ops: https://fburl.com/biggrep/h06kfrgg

masnesral replied:
@oulgen Ah thanks. So there are several. So do you think I should allowlist cond/map here specifically, i.e., because they're important? Or were those just two semi-random examples and you're advocating to allow more?

oulgen replied:
I suspect they are more popular HOPs but perhaps @zou3519 would have a more definitive answer on what is safe here

zou3519 replied:
If I'm understanding correctly, we cache compiled FX graph from inductor. This PR is saying that a graph that contains HOPs cannot be cached by default.

If so, that sounds fine to me -- caching for things like torch.cond get a bit complicated because you also want to walk through the subgraphs. cc @aakhundov

oulgen replied:
@zou3519 so Sam added a cache from FX graph to inductor generated python file. This does not work for triton HOPs because the idx on the dynamo side table needs to match which is tricky. But for torch.cond etc everything is on the graph, so I think that should be fine to cache.

masnesral replied:
@zou3519 

> Is the cache from FX GraphModule to inductor generated python file?

Exactly.

> This PR is saying that a graph that contains HOPs cannot be cached by default.

Correct. In other words, if we find find any HOPs, then bypass the cache altogether and compile normally.

> And do we recurse into the submodules of the FX GraphModule?

Hmmm, I guess I still don't know enough about inductor compilation to answer that. Maybe @eellison can help give a definite answer. Is the question about how we compute the key to look up entries or the makeup of the cache entries themselves? I don't know if a code pointer will help, but this is where the cache plugs in: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L464-L471. So at a high level, we compute the cache key from the FX GraphModule (the string representation of the code actually) and the inputs, and then we cache the output of `fx_codegen_and_compile()`

eellison replied:
torch.cond and torch.while could be supported, but they dont work "out of the box". we'd need to recurse into the subgraphs. maybe we can do that automatically, but it does not work as is today. I think it's fine to add safety in this pr and follow up and add support. 

It could also make more sense for folks who are adding the HOP to add the cache key then sam - they will know better about the details of what needs to be cached.

masnesral committed:
reuse add_kernel on "[inductor] Bypass FX graph cache when we have HigherOrderOperators"


Summary: The initial motivation was to avoid caching when we have triton higher order ops, but it's probably safer to avoid the cache for all higher order ops and allow/implement if/when we find it necessary.

Test Plan: Unit test cribbed from: https://docs-preview.pytorch.org/pytorch/tutorials/2783/recipes/torch_compile_user_defined_triton_kernel_tutorial.html?highlight=triton

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
lint on "[inductor] Bypass FX graph cache when we have HigherOrderOperators"


Summary: The initial motivation was to avoid caching when we have triton higher order ops, but it's probably safer to avoid the cache for all higher order ops and allow/implement if/when we find it necessary.

Test Plan: Unit test cribbed from: https://docs-preview.pytorch.org/pytorch/tutorials/2783/recipes/torch_compile_user_defined_triton_kernel_tutorial.html?highlight=triton

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

masnesral commented:
@pytorchbot merge
